File: plugins/capacitor-network-tools/android/src/main/java/com/t6x/plugins/networktools/NetworkToolsPlugin.kt
package com.t6x.plugins.networktools

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.*

@CapacitorPlugin(name = "NetworkTools")
class NetworkToolsPlugin : Plugin() {

    private val sshClient = SshClient()
    private val httpClient = HttpClient()
    private val tcpClient = TcpClient()
    private val udpClient = UdpClient()
    private val networkScanner = NetworkScanner()
    private val mdnsDiscovery = MdnsDiscovery()
    private val wakeOnLan = WakeOnLan()

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // === SSH ===

    @PluginMethod
    fun sshConnect(call: PluginCall) {
        val host = call.getString("host") ?: return call.reject("host is required")
        val port = call.getInt("port", 22) ?: 22
        val username = call.getString("username") ?: return call.reject("username is required")
        val password = call.getString("password")
        val privateKey = call.getString("privateKey")

        scope.launch {
            try {
                val sessionId = sshClient.connect(host, port, username, password, privateKey)
                val ret = JSObject().put("sessionId", sessionId)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("SSH connect failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun sshExec(call: PluginCall) {
        val sessionId = call.getString("sessionId") ?: return call.reject("sessionId is required")
        val command = call.getString("command") ?: return call.reject("command is required")
        val timeout = call.getInt("timeout", 30000) ?: 30000

        scope.launch {
            try {
                val result = sshClient.exec(sessionId, command, timeout.toLong())
                val ret = JSObject()
                    .put("stdout", result.stdout)
                    .put("stderr", result.stderr)
                    .put("exitCode", result.exitCode)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("SSH exec failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun sshDisconnect(call: PluginCall) {
        val sessionId = call.getString("sessionId") ?: return call.reject("sessionId is required")
        scope.launch {
            try {
                sshClient.disconnect(sessionId)
                call.resolve()
            } catch (e: Exception) {
                call.reject("SSH disconnect failed: ${e.message}", e)
            }
        }
    }

    // === SFTP ===

    @PluginMethod
    fun sftpList(call: PluginCall) {
        val sessionId = call.getString("sessionId") ?: return call.reject("sessionId is required")
        val path = call.getString("path") ?: return call.reject("path is required")

        scope.launch {
            try {
                val files = sshClient.sftpList(sessionId, path)
                val ret = JSObject().put("files", files)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("SFTP list failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun sftpDownload(call: PluginCall) {
        val sessionId = call.getString("sessionId") ?: return call.reject("sessionId is required")
        val remotePath = call.getString("remotePath") ?: return call.reject("remotePath is required")

        scope.launch {
            try {
                val data = sshClient.sftpDownload(sessionId, remotePath)
                val ret = JSObject().put("data", data)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("SFTP download failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun sftpUpload(call: PluginCall) {
        val sessionId = call.getString("sessionId") ?: return call.reject("sessionId is required")
        val remotePath = call.getString("remotePath") ?: return call.reject("remotePath is required")
        val data = call.getString("data") ?: return call.reject("data is required")

        scope.launch {
            try {
                sshClient.sftpUpload(sessionId, remotePath, data)
                call.resolve()
            } catch (e: Exception) {
                call.reject("SFTP upload failed: ${e.message}", e)
            }
        }
    }

    // === HTTP ===

    @PluginMethod
    fun httpRequest(call: PluginCall) {
        val url = call.getString("url") ?: return call.reject("url is required")
        val method = call.getString("method", "GET") ?: "GET"
        val headersObj = call.getObject("headers")
        val body = call.getString("body")
        val timeout = call.getInt("timeout", 30000) ?: 30000
        val insecure = call.getBoolean("insecure", false) ?: false

        val headers = mutableMapOf<String, String>()
        headersObj?.keys()?.forEach { key ->
            headers[key] = headersObj.getString(key) ?: ""
        }

        scope.launch {
            try {
                val result = httpClient.request(url, method, headers, body, timeout.toLong(), insecure)
                val responseHeaders = JSObject()
                result.headers.forEach { (k, v) -> responseHeaders.put(k, v) }

                val ret = JSObject()
                    .put("statusCode", result.statusCode)
                    .put("headers", responseHeaders)
                    .put("body", result.body)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("HTTP request failed: ${e.message}", e)
            }
        }
    }

    // === TCP ===

    @PluginMethod
    fun tcpConnect(call: PluginCall) {
        val host = call.getString("host") ?: return call.reject("host is required")
        val port = call.getInt("port") ?: return call.reject("port is required")
        val timeout = call.getInt("timeout", 10000) ?: 10000

        scope.launch {
            try {
                val socketId = tcpClient.connect(host, port, timeout)
                val ret = JSObject().put("socketId", socketId)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("TCP connect failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun tcpSend(call: PluginCall) {
        val socketId = call.getString("socketId") ?: return call.reject("socketId is required")
        val data = call.getString("data") ?: return call.reject("data is required")

        scope.launch {
            try {
                tcpClient.send(socketId, data)
                call.resolve()
            } catch (e: Exception) {
                call.reject("TCP send failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun tcpRead(call: PluginCall) {
        val socketId = call.getString("socketId") ?: return call.reject("socketId is required")
        val timeout = call.getInt("timeout", 10000) ?: 10000

        scope.launch {
            try {
                val data = tcpClient.read(socketId, timeout)
                val ret = JSObject().put("data", data)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("TCP read failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun tcpDisconnect(call: PluginCall) {
        val socketId = call.getString("socketId") ?: return call.reject("socketId is required")
        scope.launch {
            try {
                tcpClient.disconnect(socketId)
                call.resolve()
            } catch (e: Exception) {
                call.reject("TCP disconnect failed: ${e.message}", e)
            }
        }
    }

    // === UDP ===

    @PluginMethod
    fun udpSend(call: PluginCall) {
        val host = call.getString("host") ?: return call.reject("host is required")
        val port = call.getInt("port") ?: return call.reject("port is required")
        val data = call.getString("data") ?: return call.reject("data is required")

        scope.launch {
            try {
                udpClient.send(host, port, data)
                call.resolve()
            } catch (e: Exception) {
                call.reject("UDP send failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun udpBroadcast(call: PluginCall) {
        val port = call.getInt("port") ?: return call.reject("port is required")
        val data = call.getString("data") ?: return call.reject("data is required")

        scope.launch {
            try {
                udpClient.broadcast(port, data)
                call.resolve()
            } catch (e: Exception) {
                call.reject("UDP broadcast failed: ${e.message}", e)
            }
        }
    }

    // === Discovery ===

    @PluginMethod
    fun ping(call: PluginCall) {
        val host = call.getString("host") ?: return call.reject("host is required")
        val timeout = call.getInt("timeout", 5000) ?: 5000

        scope.launch {
            try {
                val result = networkScanner.ping(host, timeout)
                val ret = JSObject()
                    .put("reachable", result.reachable)
                    .put("latencyMs", result.latencyMs)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Ping failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun networkScan(call: PluginCall) {
        val subnet = call.getString("subnet") ?: return call.reject("subnet is required")
        val portsArray = call.getArray("ports")
        val timeout = call.getInt("timeout", 2000) ?: 2000

        val ports = if (portsArray != null) {
            (0 until portsArray.length()).map { portsArray.getInt(it) }
        } else {
            listOf(22, 80, 443, 631, 8080, 9100)
        }

        scope.launch {
            try {
                val hosts = networkScanner.scan(subnet, ports, timeout)
                val hostsArray = org.json.JSONArray()
                hosts.forEach { host ->
                    val hostObj = JSObject()
                        .put("ip", host.ip)
                        .put("hostname", host.hostname ?: "")
                        .put("openPorts", org.json.JSONArray(host.openPorts))
                        .put("latencyMs", host.latencyMs)
                    hostsArray.put(hostObj)
                }
                val ret = JSObject().put("hosts", hostsArray)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Network scan failed: ${e.message}", e)
            }
        }
    }

    @PluginMethod
    fun mdnsDiscover(call: PluginCall) {
        val serviceType = call.getString("serviceType") ?: return call.reject("serviceType is required")
        val timeout = call.getInt("timeout", 5000) ?: 5000

        scope.launch {
            try {
                val context = activity ?: return@launch call.reject("No activity context")
                val services = mdnsDiscovery.discover(context, serviceType, timeout.toLong())
                val servicesArray = org.json.JSONArray()
                services.forEach { svc ->
                    val txtRecords = JSObject()
                    svc.txtRecords.forEach { (k, v) -> txtRecords.put(k, v) }

                    val svcObj = JSObject()
                        .put("name", svc.name)
                        .put("type", svc.type)
                        .put("host", svc.host)
                        .put("port", svc.port)
                        .put("addresses", org.json.JSONArray(svc.addresses))
                        .put("txtRecords", txtRecords)
                    servicesArray.put(svcObj)
                }
                val ret = JSObject().put("services", servicesArray)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("mDNS discover failed: ${e.message}", e)
            }
        }
    }

    // === Wake-on-LAN ===

    @PluginMethod
    fun wolSend(call: PluginCall) {
        val macAddress = call.getString("macAddress") ?: return call.reject("macAddress is required")
        val broadcastAddress = call.getString("broadcastAddress", "255.255.255.255") ?: "255.255.255.255"

        scope.launch {
            try {
                wakeOnLan.send(macAddress, broadcastAddress)
                call.resolve()
            } catch (e: Exception) {
                call.reject("WoL send failed: ${e.message}", e)
            }
        }
    }

    override fun handleOnDestroy() {
        scope.cancel()
        sshClient.disconnectAll()
        tcpClient.disconnectAll()
    }
}

File: ios/Sources/NetworkToolsPlugin/NetworkToolsPlugin.swift
import Foundation
import Capacitor

@objc(NetworkToolsPlugin)
public class NetworkToolsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NetworkToolsPlugin"
    public let jsName = "NetworkTools"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "sshConnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sshExec", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sshDisconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sftpList", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sftpDownload", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sftpUpload", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "httpRequest", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "tcpConnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "tcpSend", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "tcpRead", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "tcpDisconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "udpSend", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "udpBroadcast", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "ping", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "networkScan", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "mdnsDiscover", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "wolSend", returnType: CAPPluginReturnPromise),
    ]

    private let sshClient = SshClient()
    private let httpClient = HttpClientImpl()
    private let tcpClient = TcpClientImpl()
    private let udpClient = UdpClientImpl()
    private let networkScanner = NetworkScannerImpl()
    private let mdnsDiscovery = MdnsDiscoveryImpl()
    private let wakeOnLan = WakeOnLanImpl()

    private let queue = DispatchQueue(label: "com.t6x.networktools", qos: .userInitiated, attributes: .concurrent)

    // MARK: - SSH

    @objc func sshConnect(_ call: CAPPluginCall) {
        guard let host = call.getString("host") else { return call.reject("host is required") }
        guard let username = call.getString("username") else { return call.reject("username is required") }
        let port = call.getInt("port") ?? 22
        let password = call.getString("password")
        let privateKey = call.getString("privateKey")

        queue.async {
            do {
                let sessionId = try self.sshClient.connect(host: host, port: port, username: username, password: password, privateKey: privateKey)
                call.resolve(["sessionId": sessionId])
            } catch {
                call.reject("SSH connect failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func sshExec(_ call: CAPPluginCall) {
        guard let sessionId = call.getString("sessionId") else { return call.reject("sessionId is required") }
        guard let command = call.getString("command") else { return call.reject("command is required") }
        let timeout = call.getInt("timeout") ?? 30000

        queue.async {
            do {
                let result = try self.sshClient.exec(sessionId: sessionId, command: command, timeout: timeout)
                call.resolve([
                    "stdout": result.stdout,
                    "stderr": result.stderr,
                    "exitCode": result.exitCode
                ])
            } catch {
                call.reject("SSH exec failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func sshDisconnect(_ call: CAPPluginCall) {
        guard let sessionId = call.getString("sessionId") else { return call.reject("sessionId is required") }
        queue.async {
            self.sshClient.disconnect(sessionId: sessionId)
            call.resolve()
        }
    }

    // MARK: - SFTP

    @objc func sftpList(_ call: CAPPluginCall) {
        guard let sessionId = call.getString("sessionId") else { return call.reject("sessionId is required") }
        guard let path = call.getString("path") else { return call.reject("path is required") }

        queue.async {
            do {
                let files = try self.sshClient.sftpList(sessionId: sessionId, path: path)
                call.resolve(["files": files])
            } catch {
                call.reject("SFTP list failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func sftpDownload(_ call: CAPPluginCall) {
        guard let sessionId = call.getString("sessionId") else { return call.reject("sessionId is required") }
        guard let remotePath = call.getString("remotePath") else { return call.reject("remotePath is required") }

        queue.async {
            do {
                let data = try self.sshClient.sftpDownload(sessionId: sessionId, remotePath: remotePath)
                call.resolve(["data": data])
            } catch {
                call.reject("SFTP download failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func sftpUpload(_ call: CAPPluginCall) {
        guard let sessionId = call.getString("sessionId") else { return call.reject("sessionId is required") }
        guard let remotePath = call.getString("remotePath") else { return call.reject("remotePath is required") }
        guard let data = call.getString("data") else { return call.reject("data is required") }

        queue.async {
            do {
                try self.sshClient.sftpUpload(sessionId: sessionId, remotePath: remotePath, base64Data: data)
                call.resolve()
            } catch {
                call.reject("SFTP upload failed: \(error.localizedDescription)")
            }
        }
    }

    // MARK: - HTTP

    @objc func httpRequest(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else { return call.reject("url is required") }
        let method = call.getString("method") ?? "GET"
        let headers = call.getObject("headers") as? [String: String] ?? [:]
        let body = call.getString("body")
        let timeout = call.getInt("timeout") ?? 30000
        let insecure = call.getBool("insecure") ?? false

        queue.async {
            self.httpClient.request(url: url, method: method, headers: headers, body: body, timeout: timeout, insecure: insecure) { result in
                switch result {
                case .success(let response):
                    call.resolve([
                        "statusCode": response.statusCode,
                        "headers": response.headers,
                        "body": response.body
                    ])
                case .failure(let error):
                    call.reject("HTTP request failed: \(error.localizedDescription)")
                }
            }
        }
    }

    // MARK: - TCP

    @objc func tcpConnect(_ call: CAPPluginCall) {
        guard let host = call.getString("host") else { return call.reject("host is required") }
        guard let port = call.getInt("port") else { return call.reject("port is required") }
        let timeout = call.getInt("timeout") ?? 10000

        queue.async {
            do {
                let socketId = try self.tcpClient.connect(host: host, port: port, timeout: timeout)
                call.resolve(["socketId": socketId])
            } catch {
                call.reject("TCP connect failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func tcpSend(_ call: CAPPluginCall) {
        guard let socketId = call.getString("socketId") else { return call.reject("socketId is required") }
        guard let data = call.getString("data") else { return call.reject("data is required") }

        queue.async {
            do {
                try self.tcpClient.send(socketId: socketId, base64Data: data)
                call.resolve()
            } catch {
                call.reject("TCP send failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func tcpRead(_ call: CAPPluginCall) {
        guard let socketId = call.getString("socketId") else { return call.reject("socketId is required") }
        let timeout = call.getInt("timeout") ?? 10000

        queue.async {
            do {
                let data = try self.tcpClient.read(socketId: socketId, timeout: timeout)
                call.resolve(["data": data])
            } catch {
                call.reject("TCP read failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func tcpDisconnect(_ call: CAPPluginCall) {
        guard let socketId = call.getString("socketId") else { return call.reject("socketId is required") }
        queue.async {
            self.tcpClient.disconnect(socketId: socketId)
            call.resolve()
        }
    }

    // MARK: - UDP

    @objc func udpSend(_ call: CAPPluginCall) {
        guard let host = call.getString("host") else { return call.reject("host is required") }
        guard let port = call.getInt("port") else { return call.reject("port is required") }
        guard let data = call.getString("data") else { return call.reject("data is required") }

        queue.async {
            do {
                try self.udpClient.send(host: host, port: port, base64Data: data)
                call.resolve()
            } catch {
                call.reject("UDP send failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func udpBroadcast(_ call: CAPPluginCall) {
        guard let port = call.getInt("port") else { return call.reject("port is required") }
        guard let data = call.getString("data") else { return call.reject("data is required") }

        queue.async {
            do {
                try self.udpClient.broadcast(port: port, base64Data: data)
                call.resolve()
            } catch {
                call.reject("UDP broadcast failed: \(error.localizedDescription)")
            }
        }
    }

    // MARK: - Discovery

    @objc func ping(_ call: CAPPluginCall) {
        guard let host = call.getString("host") else { return call.reject("host is required") }
        let timeout = call.getInt("timeout") ?? 5000

        queue.async {
            self.networkScanner.ping(host: host, timeout: timeout) { result in
                call.resolve([
                    "reachable": result.reachable,
                    "latencyMs": result.latencyMs
                ])
            }
        }
    }

    @objc func networkScan(_ call: CAPPluginCall) {
        guard let subnet = call.getString("subnet") else { return call.reject("subnet is required") }
        let ports = (call.getArray("ports") as? [Int]) ?? [22, 80, 443, 631, 8080, 9100]
        let timeout = call.getInt("timeout") ?? 2000

        queue.async {
            self.networkScanner.scan(subnet: subnet, ports: ports, timeout: timeout) { hosts in
                let hostsArray = hosts.map { host -> [String: Any] in
                    return [
                        "ip": host.ip,
                        "hostname": host.hostname ?? "",
                        "openPorts": host.openPorts,
                        "latencyMs": host.latencyMs
                    ]
                }
                call.resolve(["hosts": hostsArray])
            }
        }
    }

    @objc func mdnsDiscover(_ call: CAPPluginCall) {
        guard let serviceType = call.getString("serviceType") else { return call.reject("serviceType is required") }
        let timeout = call.getInt("timeout") ?? 5000

        queue.async {
            self.mdnsDiscovery.discover(serviceType: serviceType, timeout: timeout) { services in
                let servicesArray = services.map { svc -> [String: Any] in
                    return [
                        "name": svc.name,
                        "type": svc.type,
                        "host": svc.host,
                        "port": svc.port,
                        "addresses": svc.addresses,
                        "txtRecords": svc.txtRecords
                    ]
                }
                call.resolve(["services": servicesArray])
            }
        }
    }

    // MARK: - Wake-on-LAN

    @objc func wolSend(_ call: CAPPluginCall) {
        guard let macAddress = call.getString("macAddress") else { return call.reject("macAddress is required") }
        let broadcastAddress = call.getString("broadcastAddress") ?? "255.255.255.255"

        queue.async {
            do {
                try self.wakeOnLan.send(macAddress: macAddress, broadcastAddress: broadcastAddress)
                call.resolve()
            } catch {
                call.reject("WoL send failed: \(error.localizedDescription)")
            }
        }
    }
}
